spec/index.adoc

```diff
@@ -4109,14 +4126,15 @@ callbacks = λ(callbacks, response, refunded_cycles, sysenv, available) → λ w
     if typeof(func) ≠ func (i32) -> () then Trap
 
     let es = ref { empty_execution_state with
-      wasm_state;
+      wasm_state = wasm_state;
     }
     func<es>(callbacks.on_cleanup.env)
     Return {
```

Comment:
Interface Spec: "During the execution of the cleanup function, only a subset of the System API is available (namely ic0.debug_print, ic0.trap and the ic0.stable_* functions)."
However, we do not trap here if some of the functions have actually been invoked, e.g., for query methods, we perform the following checks:
```
  if es.cycles_accepted ≠ 0 then Trap {cycles_used = es.cycles_used;}
  if es.calls ≠ () then Trap {cycles_used = es.cycles_used;}
  if es.ingress_filter ≠ Reject then Trap {cycles_used = es.cycles_used;}
  if es.response = NoResponse then Trap {cycles_used = es.cycles_used;}
```

Reply:
Good catch, probably simply missing.

(I am not sure if this style of “trap at the end if an unwanted getter/setter was used” is really helpful, compared to being more explicit about the context and trapping in the function directly…)

Reply:
I've pushed a bunch of if's analogously by the query case (except that for callback cleanup, there can be no reply). I've added a note in my TODO list to handle trapping on illegal operations according to `spec/ic0.txt`, but let's do it outside this PR.